pkg/controllers/follower/controller.go

```diff
@@ -425,113 +450,104 @@ func (c *Controller) updateFollower(
 		return worker.StatusAllOK
 	}
 
-	followerObj := &fedtypesv1a1.GenericFederatedFollower{}
-	err = runtime.DefaultUnstructuredConverter.FromUnstructured(followerUns.Object, followerObj)
-	if err != nil {
-		logger.Error(err, "Failed to unmarshall follower object from unstructured")
-		return worker.StatusAllOK // retrying won't help
-	}
-
-	currentLeaders := sets.New(followerObj.Spec.Follows...)
+	currentLeaders := sets.New(followerObject.GetSpec().Follows...)
 	c.cacheObservedFromFollowers.update(follower, currentLeaders)
 
 	desiredLeaders := c.cacheObservedFromLeaders.reverseLookup(follower)
 
 	leadersChanged := !equality.Semantic.DeepEqual(desiredLeaders, currentLeaders)
 	updated, err := c.updateFollower(
 		ctx,
-		handles,
-		followerUns,
-		followerObj,
+		followerObject,
 		leadersChanged,
 		desiredLeaders.UnsortedList(),
 	)
 	if err != nil {
 		if apierrors.IsConflict(err) {
 			return worker.StatusConflict
 		}
-		logger.Error(err, "Failed to update follower")
+		keyedLogger.Error(err, "Failed to update follower")
 		c.eventRecorder.Eventf(
-			followerUns,
+			followerObject,
 			corev1.EventTypeWarning,
 			EventReasonFailedUpdateFollower,
 			"Failed to update follower to sync with leader placements: %v",
 			err,
 		)
 		return worker.StatusError
 	} else if updated {
-		logger.V(1).Info("Updated follower to sync with leaders")
+		keyedLogger.V(1).Info("Updated follower to sync with leaders")
 	}
 
 	return worker.StatusAllOK
 }
 
 func (c *Controller) getLeaderObj(
-	leader fedtypesv1a1.LeaderReference,
-) (*typeHandles, *fedtypesv1a1.GenericObjectWithPlacements, error) {
+	leader fedcorev1a1.LeaderReference,
+) (*typeHandles, fedcorev1a1.GenericFederatedObject, error) {
 	leaderGK := leader.GroupKind()
 	handles, exists := c.leaderTypeHandles[leaderGK]
 	if !exists {
 		return nil, nil, fmt.Errorf("unsupported leader type %v", leaderGK)
 	}
-	leaderQualifiedName := common.QualifiedName{Namespace: leader.Namespace, Name: leader.Name}
-	leaderUns, err := getObjectFromStore(
-		handles.informer.Informer().GetStore(),
-		leaderQualifiedName.String(),
+
+	leaderName := naming.GenerateFederatedObjectName(leader.Name, handles.typeConfig.Name)
+	leaderObj, err := fedobjectadapters.GetFromLister(
+		c.fedObjectInformer.Lister(),
+		c.clusterFedObjectInformer.Lister(),
+		leader.Namespace,
+		leaderName,
 	)
-	if err != nil {
+	if err != nil && !apierrors.IsNotFound(err) {
 		return nil, nil, fmt.Errorf("get from store: %w", err)
 	}
-	if leaderUns == nil {
+	if apierrors.IsNotFound(err) {
 		return nil, nil, nil
 	}
 
-	leaderObj, err := util.UnmarshalGenericPlacements(leaderUns)
-	if err != nil {
-		return nil, nil, fmt.Errorf("unmarshal to generic object with placements: %w", err)
-	}
-
 	return handles, leaderObj, nil
 }
 
 func (c *Controller) leaderPlacementUnion(
-	leaders []fedtypesv1a1.LeaderReference,
+	leaders []fedcorev1a1.LeaderReference,
 ) (map[string]struct{}, error) {
 	clusters := map[string]struct{}{}
```

Comment:
nit: we could make `clusters` a `sets.Set[string]`. This would simplify the calculation of leader placement union, and conversion of cluster names to a slice.